### PR TITLE
Removed Check Invalid Path Chars

### LINKED ISSUE #25450: Should remove CheckInvalidPathChars
The check for invalid path chars on the majority of System.IO APIs aggregates to a significant cost in real world performance metrics. There is little benefit to premptively checking- we should start by removing the check from everything but Path.GetFullPath().

http://source.dot.net/#System.Private.CoreLib/shared/System/IO/PathInternal.cs,486d9b8f9e6ecd64,references

We need to remove checks in CoreCLR (where Path lives) and System.IO.FileSystem.

### PR DESCRIPTION
**Pull request by Anipik**
Fixes #25450 
coreclr pr - dotnet/coreclr#15191

### CODE REVIEW

File: src/System.IO.FileSystem/src/System/IO/Win32FileSystemEnumerable.cs [hunk 1]
```diff
@@ -217,7 +217,6 @@ private Win32FileSystemEnumerableIterator(string fullPath, string normalizedSear
 
             if (searchCriteria != null)
             {
-                PathInternal.CheckInvalidPathChars(fullPath);
```
Comment (danmosemsft) [inline src/System.IO.FileSystem/src/System/IO/Win32FileSystemEnumerable.cs]: this is not the only use in corefx. remove the impl and test for checkinvalidpathchars as well.

**Comment by danmosemsft** [comment]
LGTM, @JeremyKuhne can sign off.

File: src/Common/src/System/IO/PathInternal.cs [hunk 2]
```diff
@@ -98,8 +82,7 @@ internal static StringBuilder TrimEnd(this StringBuilder builder, params char[]
         internal static int FindFileNameIndex(string path)
```
Comment (stephentoub) [inline src/Common/src/System/IO/PathInternal.cs]: Where is this method used?  Can it just be deleted?

File: src/Common/src/System/IO/PathInternal.cs [hunk 3]
```diff
@@ -10,22 +10,6 @@ namespace System.IO
...
-        /// </summary>
-        /// <exception cref="System.ArgumentNullException">Thrown if the path is null.</exception>
-        /// <exception cref="System.ArgumentException">Thrown if the path has invalid characters.</exception>
-        /// <param name="path">The path to check for invalid characters.</param>
-        internal static void CheckInvalidPathChars(string path)
-        {
-            if (path == null)
-                throw new ArgumentNullException(nameof(path));
-
-            if (HasIllegalCharacters(path))
```
Comment (stephentoub) [inline src/Common/src/System/IO/PathInternal.cs]: @JeremyKuhne, what about direct call sites to HasIllegalCharacters, e.g. https://github.com/dotnet/corefx/blob/0eb5e7451028e9374b8bb03972aa945c128193e1/src/System.IO.FileSystem.DriveInfo/src/System/IO/DriveInfo.Windows.cs#L25-L26 ?

Comment (JeremyKuhne) [reply src/Common/src/System/IO/PathInternal.cs]: These should be removed too.

File: src/Common/src/System/IO/PathInternal.cs (hunk 2 above)
Comment (Anipik) [reply src/Common/src/System/IO/PathInternal.cs]: @JeremyKuhne  should I delete this too ?

Comment (JeremyKuhne) [reply src/Common/src/System/IO/PathInternal.cs]: If it isn't used anywhere go ahead and delete the method. I _think_ it is only used in CoreCLR.

File: src/Common/src/System/IO/PathInternal.cs (hunk 3 above)
Comment (Anipik) [reply src/Common/src/System/IO/PathInternal.cs]: should I remove the function  HasIllegalCharacters  from the corefx as it is not being used anywhere else

Comment (JeremyKuhne) [reply src/Common/src/System/IO/PathInternal.cs]: Yes, please go ahead and remove any dead methods.

Comment (Anipik) [reply src/Common/src/System/IO/PathInternal.cs]: <test name="System.IO.FileSystem.DriveInfoTests.DriveInfoWindowsTests.Ctor_InvalidPath_ThrowsArgumentException(driveName: \&quot;:\\0\&quot;, paramName: \&quot;driveName\&quot;)" type="System.IO.FileSystem.DriveInfoTests.DriveInfoWindowsTests" method="Ctor_InvalidPath_ThrowsArgumentException" time="0.0033929" result="Fail">
        <failure exception-type="Xunit.Sdk.EqualException">
          <message><![CDATA[Assert.Equal() Failure\r\n          ↓ (pos 0)\r\nExpected: driveName\r\nActual:   path\r\n          ↑ (pos 0)]]></message>
          <stack-trace><CDATA[   at System.AssertExtensions.Throws[T in c:\git\corefx3\src\CoreFx.Private.TestUtilities\src\System\AssertExtensions.cs:line 39
   at System.IO.FileSystem.DriveInfoTests.DriveInfoWindowsTests.Ctor_InvalidPath_ThrowsArgumentException(String driveName, String paramName) in c:\git\corefx3\src\System.IO.FileSystem.DriveInfo\tests\DriveInfo.Windows.Tests.cs:line 36]]></stack-trace>
        </failure>
      </test>

@JeremyKuhne 
this is the disabled test that failed after removing the check for hasIllegalCharacters. should I remove\disable the test or keep this check.

Comment (JeremyKuhne) [reply src/Common/src/System/IO/PathInternal.cs]: You can remove the ones that hit the invalid char checks. There is also a check for proper formatting of the `C:`, so some of these are still valid.

**Pull request merged**